homeassistant/components/madvr/config_flow.py

```diff
@@ -0,0 +1,174 @@
+"""Config flow for the integration."""
+
+import asyncio
+import logging
+from typing import Any
+
+import aiohttp
+from madvr.madvr import HeartBeatError, Madvr
+import voluptuous as vol
+
+from homeassistant.config_entries import (
+    ConfigEntry,
+    ConfigFlow,
+    ConfigFlowResult,
+    OptionsFlow,
+)
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_NAME, CONF_PORT
+from homeassistant.core import callback
+
+from .const import CONF_POWER_ON, DEFAULT_NAME, DEFAULT_PORT, DOMAIN
+from .coordinator import MadVRCoordinator
+from .utils import CannotConnect
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(
+            CONF_HOST,
+        ): str,
+        vol.Required(
+            CONF_MAC,
+        ): str,
+        vol.Required(
+            CONF_PORT,
+            default=DEFAULT_PORT,
+        ): int,
+        vol.Required(
+            CONF_POWER_ON,
+            default=False,
+        ): bool,
+    }
+)
+
+
+class MadVRConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for the integration."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            port = user_input[CONF_PORT]
+            mac = user_input[CONF_MAC]
+            keep_power_on = user_input[CONF_POWER_ON]
+            try:
+                await self._test_connection(host, port, mac, keep_power_on)
+                return self.async_create_entry(
+                    title=user_input.pop(CONF_NAME, DEFAULT_NAME),
+                    data=user_input,
+                )
+            except CannotConnect:
+                _LOGGER.error("CannotConnect error caught")
+                errors["base"] = "cannot_connect"
+        # Whether it's the first attempt or a retry, show the form
+        return self.async_show_form(
+            step_id="user",
+            data_schema=self.add_suggested_values_to_schema(
+                STEP_USER_DATA_SCHEMA, user_input
+            ),
+            errors=errors,
+        )
+
+    async def _test_connection(
+        self, host: str, port: int, mac: str, keep_power_on: bool
+    ):
+        """Test if we can connect to the device."""
+        try:
+            madvr_client = Madvr(host=host, port=port, mac=mac)
+            _LOGGER.debug(
+                "Testing connection to MadVR at %s:%s with mac %s", host, port, mac
+            )
+            # turn on the device
+            await madvr_client.power_on()
```

Comment:
Does this now still use Wake on Lan? We should not do that, its not a strange requirement to have people turn on their device before being able to add it

Reply:
wol removed

Reply:
How doe turning on now work?

Reply:
without wake on lan, it is not possible. It must have wake on lan implemented to turn it on

Reply:
So you're saying you still use wake on Lan?

Reply:
no if a mac is not provided the library will not do anything when the power command is called. It is impossible

Reply:
But then why do we implement something that does not work?

Reply:
it doesnt work because you wont let me include wake on lan inside the library :)

it will work once device trigger is implemented, but Im trying to tell you it is _impossible_ to action this device without wake on lan and not being able to call power on is a confusing user experience

Reply:
But then we should not implement `async_turn_on` if we can't turn it on via the library and not just implement a library call that doesn't work

Reply:
okay I will remove for now then

Reply:
actually a counter proposition

The original complaint was entering a mac is bad. So how about during initial config, asking the user to turn it on and connecting, then I can make a call to the device to get the mac and store it, then use wake on lan (inside the library) to control it

This seems like best of both worlds and letting the device handle the power lifecycle and also not asking user for mac input